### ISSUE TITLE
[MNG-5896] Download dependency POMs in parallel

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7539]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** opened **[MNG-5896](https://issues.apache.org/jira/browse/MNG-5896?redirect=false)** and commented

### Background

When building a project with dependencies not yet available in the local repository, I noticed that Maven 3.3.3 first downloads the dependency POMs _sequentially_ and then proceeds downloading the dependency JARs with up to 5 threads _in parallel_.

Due to this, when first building a project with a large number of dependencies, downloading a large number of small POMs may take a lot longer than downloading the much larger JARs, or even longer than building the project itself, especially when a repository manager is used which increases the download latency.

### Enhancement

Download POMs of (transitive) dependencies in parallel to significantly speed up initial builds of large projects.

---

**Affects:** 3.3.3

**Issue Links:**
- [MRESOLVER-7](https://issues.apache.org/jira/browse/MRESOLVER-7) Download dependency POMs in parallel in BF collector
 (_**"is blocked by"**_)
- [MNG-6110](https://issues.apache.org/jira/browse/MNG-6110) Upgrade Aether to Maven Resolver
 (_**"is blocked by"**_)
- [MNG-7677](https://issues.apache.org/jira/browse/MNG-7677) Maven 3.9.0 is ~10% slower than 3.8.7 in large multi-module builds

- [MRESOLVER-324](https://issues.apache.org/jira/browse/MRESOLVER-324) Make the BF algorithm as the default option to speed up maven dependency resolution and downloading

- [MNG-6672](https://issues.apache.org/jira/browse/MNG-6672) Upgrade Maven Resolver to 1.4.2
 (_**"requires"**_)

10 votes, 22 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You should try to build Nexus, this project requires time. I would be curious to see the performance gain.

**Comment by jira-importer** [comment]
**[Testo Nakada](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=testn)** commented

This issue should be a low hanging fruit to help improve the build time.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann), can you revise your patch for Maven Resolver? Aether has been imported back to us. I am willing to review your patch and merge it into Maven Resolver 1.2.

**Comment by jira-importer** [comment]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** commented

Sure, I've been waiting all the time for Aether to be integrated from Eclipse.

Could you give some more details - is Maven Resolver a separate plugin or project, and where do I find it? What would be the  Maven target release and the appropriate branch for testing this?

**Comment by jira-importer** [comment]
**[Harald Wellmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hwellmann)** commented

I've reworked my patch, created a new issue MRESOLVER-7 and submitted a pull request.

I also did a local build of 3.4.0-SNAPSHOT, replacing the Aether dependencies by Maven Resolver. Is there already a branch for that dependency upgrade, or should I submit another pull request for that?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Yes, please. Create a Maven issue: "Upgrade Aether to Maven Resolver 1.2" and create the PR. I intend to roll Maven 3.4 with our new Resolver.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This will be resolved as soon as Maven Resolver 1.4.0 is in use.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with MNG-6672

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

reverted in https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=383c8c11ef793025db6107fbfeb7ccd5fdd9ba87 because MRESOLVER-7 was causing infinite loops

**Comment by jira-importer** [comment]
**[Fred Bricon](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fbricon)** commented

Now that https://github.com/apache/maven-resolver/pull/178 has been merged, can this issue be closed? Which version of Maven will include it?

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Fred Bricon](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fbricon) 

This parallel pom downloading happens in a brand new BF collector as of now. BF collector is a combination of 
https://github.com/apache/maven-resolver/pull/158
https://github.com/apache/maven-resolver/pull/178

It includes some performance improvements based on what we've done in our company to speed up maven build. So far the BF and original DF algorithm co-exists and the DF is still the default.

When maven 3.9.x ships with maven-resolver 1.9.0, then users can simply put a [.mvn/maven.config](https://maven.apache.org/configure.html) within project root with below content:
"-Daether.dependencyCollector.impl=bf" to enable these features:
* Skipper to skip unnecessary resolution
* Parallel descriptor files (pom, metadata.xml) downloading

Not sure what's the plan to make the BF as default one. I believe this JIRA can be closed accordingly until BF becomes the default option.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) 

May I know the additional things I can help to make BF the default choice. More tests? Or have to wait for more user feedback?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai), we will first integrate this into 3.9.0 and 4.0.0-alpha-3 to have user feedback (negative/positive). If there is no negative one then we can discuss a change.

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai) shouldn't it be "-Daether.collector.impl=bf"

Based on: https://github.com/apache/maven-resolver/blob/7fdaabceae2633ad63525bb980f57d266eed88cd/maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/collect/DefaultDependencyCollector.java#L50

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber) You are right.

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) 

Not sure you would prefer updating doc to "aether.collector.impl" or updating code to "aether.dependencyCollector.impl".

Also looks like 1.9.0 was not released successfully? I saw the doc: https://maven.apache.org/resolver/configuration.html still saying latest version is 1.8.2.

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

I think updating the code to  "aether.dependencyCollector.impl" makes sense, since other config options documented here have "aether.dependencyCollector" prefix: https://maven.apache.org/resolver/configuration.html

I built a maven 3.9.0 tarball with maven resolver 1.9.0 (defaulting to BF instead of DF) and the performance looks great so far. I'll roll this out to more of our engineering team and see if we run into any issues.

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber) [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) [Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas) 

Thanks for feedback.
Hope we can hear more updates from your team in near future.

Fixed the param in https://github.com/apache/maven-resolver/pull/217

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

On one of our medium-large projects (24 modules, ~10,000 total dependencies) we're seeing mvn dependency:resolve take 111 seconds using DF impl, and 15 seconds using BF impl. And this is with the offline option. When you actually need to download/update artifacts, that gap only widens.

No issues with correctness or stability reported yet (using BF + 10 threads)

EDIT: I thought that using the DF option would perform similarly to previous versions of Maven, and that this would be a mostly valid performance comparison. However, when I run the same command on the same repo using Maven 3.8.6, it also finishes in 15 seconds. Which is confusing, because I thought Maven 3.8.6 is effectively using the df strategy?

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber) 

Should not be such case. **DF** in 3.9.0 should not have too much changes causing such big performance downgrade.

Could you share the detailed steps and some test project to reproduce such issue?

**Comment by jira-importer** [comment]
**[wei cai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wecai)** commented

[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber) 

Tried to build a maven 3.9.0 snapshot myself, and did not see issues regarding with DF. 
Have you already found the cause?

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

Sorry for the lengthy delay, I put together a repository that I believe reproduces the issue:

https://github.com/jhaber/maven-resolver-test

There are some directions in the readme. I also uploaded the full Maven folder to the repo in case it's an issue with how I built Maven 3.9.0 from source (I cloned https://github.com/apache/maven, checked out maven-3.9.x branch, and ran 'mvn -DdistributionTargetDir="$HOME/apache-maven-3.9.0-SNAPSHOT" clean package')

On my laptop, Maven 3.8.7 can run dependency:resolve in ~16 seconds. Maven 3.9.0 (with maven-resolver 1.9.4) takes ~31 seconds to run the same command. Running Maven 3.9.0 with BF mode brings the performance back to ~16 seconds.

I haven't dug into the root cause, but I would expect Maven 3.9.0 in DF mode to take ~16 seconds (ie, same performance as Maven 3.8.7), and BF mode to improve upon that.

**Comment by jira-importer** [comment]
**[Jonathan Haber](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jhaber)** commented

My hunch is that this performance regression is related to https://issues.apache.org/jira/browse/MNG-7677 and not related to the BF/DF changes.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This is not a dup of MNG-7677.
